lib/ecto/repo/assoc.ex

```diff
@@ -37,9 +37,13 @@ defmodule Ecto.Repo.Assoc do
   defp merge([struct|sub_structs], {keys, dict, sub_dicts}, parent_key) do
     child_key =
       if struct do
-        [{_, key}] = Ecto.primary_key!(struct)
-        key || raise Ecto.NoPrimaryKeyValueError, struct: struct
+        Ecto.primary_key!(struct)
+        |> Enum.reduce([], fn(x, acc) ->
+          {_, key} = x
+          [key | acc]
+        end)
       end
+    if Enum.empty?(child_key), do: raise Ecto.NoPrimaryKeyValueError, struct: struct
```

Comment:
Yes, I think so. It doesn't make sense to have `NULL` in one of the parts of a composite key, does it?